pkg/devfile/adapters/common/command_supervisor.go

```diff
@@ -0,0 +1,60 @@
+package common
+
+import (
+	"fmt"
+	"github.com/openshift/odo/pkg/devfile/parser/data/common"
+	"github.com/openshift/odo/pkg/machineoutput"
+)
+
+// supervisorCommand encapsulates a supervisor-specific command
+type supervisorCommand struct {
+	adapter commandExecutor
+	cmd     []string
+	info    ComponentInfo
+}
+
+// newSupervisorInitCommand creates a command that initializes the supervisor for the specified devfile if needed
+// nil is returned if no devfile-specified container needing supervisor initialization is found
+func newSupervisorInitCommand(command common.DevfileCommand, adapter commandExecutor) (command, error) {
+	cmd := []string{SupervisordBinaryPath, "-c", SupervisordConfFile, "-d"}
+	info, err := adapter.SupervisorComponentInfo(command)
+	if err != nil {
+		adapter.Logger().ReportError(err, machineoutput.TimestampNow())
+		return nil, err
+	}
+	if !info.IsEmpty() {
+		return supervisorCommand{
+			adapter: adapter,
+			cmd:     cmd,
+			info:    info,
+		}, nil
+	}
+	return nil, nil
+}
+
+// newSupervisorStopCommand creates a command implementation that stops the specified command via the supervisor
+func newSupervisorStopCommand(command common.DevfileCommand, executor commandExecutor) (command, error) {
+	cmd := []string{SupervisordBinaryPath, SupervisordCtlSubCommand, "stop", "all"}
+	if stop, err := newOverriddenSimpleCommand(command, executor, cmd); err == nil {
+		// override spinner message
+		stop.msg = fmt.Sprintf("Stopping %s command %q, if running", command.GetID(), command.Exec.CommandLine)
```

Comment:
I think this adds a extra line to the push output
```
Executing devfile commands for component nodejs
 ✓  Executing install command "npm install", if not running [2s]
 ✓  Stopping run command "npm start", if running [1s]
 ✓  Executing run command "npm start", if not running [1s]
```
Is this intentional?

Reply:
Yes. Unless we want to encapsulate both as a `Restarting…` log but that would complicate the code since the output would need to be computed from several commands instead of outputting the result of each command. Is this a problem?

Reply:
It feels kind of complicated and more verbose to me. The earlier output
```
Executing devfile commands for component nodejs
 ✓  Executing install command "npm install" [6s]
 ✓  Executing run command "npm start" [1s]
```
was simpler. I don't think the user needs to know about the restarting part. They can always do `-v 4` for more details.

Reply:
Fixed…